### PR TITLE
Initial implementation of MimeToken

### PR DESCRIPTION
**Pull request by 0xGabi**

### CODE REVIEW

**Comment by PJColombo** [comment]
Amazing work @0xGabi!

I think it would be interesting to keep the ERC20 compatibility.

We could implement the missing ERC20 functions with a `revert`:

- [transfer(to, amount)](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20-transfer-address-uint256-)

- [allowance(owner, spender)](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20-allowance-address-address-)

- [approve(spender, amount)](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20-approve-address-uint256-)

- [transferFrom(from, to, amount)](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20-transferFrom-address-address-uint256-)

**Review by PJColombo — COMMENTED** [review]
Should we make it upgradeable?

File: src/MimeToken.sol [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        _claimedBitMapAt[_currentRound][claimedWordIndex] =
+            _claimedBitMapAt[_currentRound][claimedWordIndex] | (1 << claimedBitIndex);
+    }
+
+    function claim(uint256 index, address account, uint256 amount, bytes32[] calldata merkleProof) public {
+        if (isClaimed(index)) revert AlreadyClaimed();
+
+        // Verify the merkle proof.
+        bytes32 node = keccak256(abi.encodePacked(index, account, amount));
+        if (!MerkleProof.verify(merkleProof, _merkleRootAt[_currentRound], node)) revert InvalidProof();
```
Comment (PJColombo) [inline src/MimeToken.sol]:
```suggestion
        if (!MerkleProof.verify(merkleProof, merkleRoot(), node)) revert InvalidProof();
```

**Comment by 0xGabi** [comment]
I'm not sure if we want the token to be upgradeable. Instead we can allow pool owners to update the token. What do you think?

**Review by PJColombo — APPROVED** [review]
LGTM

**Comment by PJColombo** [comment]
> I'm not sure if we want the token to be upgradeable. Instead we can allow pool owners to update the token. What do you think?

Yes, we can do something along the lines of what we did with the osmotic pools: use the beacon proxy pattern and transfer the owership to the pool owners

**Pull request merged**